test/hipcub/test_hipcub_vector.cpp

```diff
@@ -61,11 +61,11 @@ TYPED_TEST_SUITE(HipcubVector, Params);
 
 template<class T, unsigned int VecSize, unsigned int BlockSize>
 __global__
-void vector_double_kernel(T* device_input, T* device_output)
+void vector_test_kernel(T* device_input, T* device_output)
 {
     unsigned int index = hipThreadIdx_x + (hipBlockIdx_x * BlockSize);
 
-    device_output[index] = device_input[index] + device_input[index];
+    device_output[index] = device_input[index] - device_input[index];
```

Comment:
I would add a comment explaining why negation is used here.

```suggestion
    // On the device (only on Windows), the addition of bools (which is non-zero) is not necessarily 1, even after writing back to the output bool pointer. So bitwise comparison with the result on host can fail. We therefore use subtraction instead of addition.
    device_output[index] = device_input[index] - device_input[index];
```

Reply:
I dug into this problem further, and it seems that the root cause is that `hipcub/backend/rocprim/util_type.hpp` defines the `CubVector` type for bools using `unsigned char` as the backing storage type:
`HIPCUB_DEFINE_VECTOR_TYPE(bool, unsigned char)`

The add that's performed on the device in `vector_double_kernel` operates on `unsigned char`, and it doesn't actually cast back to `bool` there - the result is stored as an `unsigned char` in the `device_output[index]` vector:
`device_output[index] = device_input[index] - device_input[index];`

It's only actually cast back to a `bool` later on the host, and it's done using a `reinterpret_cast` that operates on the pointer-versions of the types:
`ASSERT_NO_FATAL_FAILURE(
                test_utils::assert_eq(*(reinterpret_cast<const T*>(&output_vec) + j),
                                      expected_num));`

Because we're `reinterpret_cast`ing a pointer (and then dereferencing), the compiler doesn't enforce the C++ conversion rules for `uchar` -> `bool` (it doesn't convert non-zero values to 1).

If I instead do the cast in two steps (first `reinterpret_cast` to the storage type (`unsigned char`), and then `static_cast` each element to `bool`), then it works, and the test passes:
`using StorageType = typename hipcub::CubVector<bool, vec_size>::value_type;`
`StorageType storage_val = *(reinterpret_cast<const StorageType*>(&output_vec) + j);`
`bool bval = static_cast<bool>(storage_val);`
`ASSERT_NO_FATAL_FAILURE(test_utils::assert_eq(bval, expected_num));`

Unfortunately, there's no platform-independent way to grab the actual storage type. On HIP, it's value_type, and on CUB/CCCL, it's BaseType. We should probably log a bug with the HIP runtime team about this.

Maybe I'll stick with using subtraction for now, but add a different note in the comments to reflect the actual root cause.

Reply:
I've updated the comment.
I'll log a runtime ticket separately that requests a platform-independent way to grab the actual storage type. That should allow us to fix the root cause going forward.